service/s3/endpoints.go

```diff
@@ -213,6 +215,127 @@ func finalizeClientEndpointResolverOptions(options *Options) {
 
 }
 
+type legacyEndpointResolverAdapter struct {
+	legacyResolver EndpointResolver
+	resolver       EndpointResolverV2
+}
+
+func (l *legacyEndpointResolverAdapter) ResolveEndpoint(ctx context.Context, params EndpointParameters) (endpoint smithyendpoints.Endpoint, err error) {
+	var fips aws.FIPSEndpointState
+	var dualStack aws.DualStackEndpointState
+
+	if aws.ToBool(params.UseFIPS) {
+		fips = aws.FIPSEndpointStateEnabled
+	}
+	if aws.ToBool(params.UseDualStack) {
+		dualStack = aws.DualStackEndpointStateEnabled
+	}
+
+	resolveEndpoint, err := l.legacyResolver.ResolveEndpoint(aws.ToString(params.Region), EndpointResolverOptions{
+		ResolvedRegion:       aws.ToString(params.Region),
+		UseFIPSEndpoint:      fips,
+		UseDualStackEndpoint: dualStack,
+	})
+	if err != nil {
+		return endpoint, err
+	}
+
+	if resolveEndpoint.HostnameImmutable {
+		uriString := resolveEndpoint.URL
+		uri, err := url.Parse(uriString)
+		if err != nil {
+			return endpoint, fmt.Errorf("Failed to parse uri: %s", uriString)
+		}
+
+		return smithyendpoints.Endpoint{
+			URI: *uri,
+		}, nil
+	}
+
+	if resolveEndpoint.Source == aws.EndpointSourceServiceMetadata {
+		return l.resolver.ResolveEndpoint(ctx, params)
+	}
+
+	params.UseDualStack = nil
```

Comment:
A few questions:
1. Shouldn't these booleans be set to their defaults instead?
2. How are we going to handle service-specific (`clientContextParam`) and operation-specific (`contextParam` / `staticContextParam`) bindings in this adapter?

Reply:
For (2) aren't the params already bound?

Reply:
1. yah i could see it going either way.  i think setting them to `nil` signals that this input was already used when resolving a URL. but i can make this change for now
2. its handled a layer above in the middleware. i can explain this in more depth offline

Reply:
Yeah you're right on (2) - I missed that it was being passed into this function.